src/codec.rs

```diff
@@ -1498,4 +1595,61 @@ mod tests {
 		assert_eq!(MyWrapper(3u32.into()).encode(), result);
 		assert_eq!(MyWrapper::decode(&mut &*result).unwrap(), MyWrapper(3_u32.into()));
 	}
+
+	macro_rules! check_bound {
+		( $m:expr, $ty:ty, $typ1:ty, [ $($ty2:ty),* ]) => {
+			$(
+				check_bound!($m, $ty, $typ1, $ty2);
+			)*
+		};
+		( $m:expr, $ty:ty, $typ1:ty, $ty2:ty) => {
+			let enc = ((<$ty>::max_value() >> 2) as $typ1 << 2) | $m;
+			assert!(Compact::<$ty2>::decode(&mut &enc.to_le_bytes()[..]).is_err());
+		};
+	}
+	macro_rules! check_bound_u32 {
+		( [ $($ty2:ty),* ]) => {
+			$(
+				check_bound_u32!($ty2);
+			)*
+		};
+		( $ty2:ty) => {
+			let mut dest = Vec::new();
+			dest.push(0b11);
+			for _ in 0..3 {
+				dest.push(u8::max_value());
+			}
+			dest.push(u8::max_value() >> 2);
+			assert!(Compact::<$ty2>::decode(&mut &[0b11, 0xff, 0xff, 0xff, 0xff >> 2][..]).is_err());
+
+		};
+	}
+	macro_rules! check_bound_high {
+		( $m:expr, [ $($ty2:ty),* ]) => {
+			$(
+				check_bound_high!($m, $ty2);
+			)*
+		};
+		( $s:expr, $ty2:ty) => {
+			let mut dest = Vec::new();
+			dest.push(0b11 + (($s - 4) << 2) as u8);
+			for _ in 0..($s - 1) {
+				dest.push(u8::max_value());
+			}
+			dest.push(0);
+			assert!(Compact::<$ty2>::decode(&mut &dest[..]).is_err());
```

Comment:
Please check the exact error message, your test would be successful even if some other error would be returned.